### DISCUSSION TITLE
Ebyte E22-400M30S

### DISCUSSION THREAD

**Discussion by EliasA97** [jgromes/RadioLib#415] [Q&A] [answered]
Hello, I'm designing a custom board with atmega 2560 and the ebyte module from the title and I have some questions:
 I know in fact that this module needs 3.3V for the communication input pins (RxEn, TxEn, DIO1, DIO2 , NRST, MOSI, SCK, NSS). The BUSY and MISO are outputs as I read from the manual. The DIO1 and DIO2 are input-output. What are their use? Does the library use them as inputs or outputs? Do I have to get them a voltage level conversion as the other input pins? Also what about the DIO2 and TxEn. Do I have to connect them together or isn't needed?
Thank's!

**Comment by jgromes** [comment, marked as answer]
The primary function of all DIOx pins is to provide digital signals to or from the SX126x. In that they can all be used interchangeably. However, DIO2 and DIO3 also have special functions. DIO2 can act as an RF switch control pin, while DIO3 can act as voltage reference for TCXO. By default, that's the configuration RadioLib uses. DIO1 is then used as a source of external interrupts for the microcontroller.

So, if you are going to use the default configuration, DIO1 is an output of the module, while DIO2 is not connected to the microcontroller. E22-400M30S has an RF switch that you have to control using the RxEN and TxEN pins. You could connect RxEN and TxEN to some GPIO pins of the ATmega (and let RadioLib handle the switching), but it's easier to let SX126x handle the switching on its own using DIO2 with a simple transistor inverter (Q3). Q4 isn't really needed, it's just to power up an LED whenever the module is transmitting.

The following is taken from my design: https://github.com/jgromes/ESP32-1W-GroundStation

![Screenshot_19](https://user-images.githubusercontent.com/3407192/144375844-85f34906-0829-4542-a7a1-8474c0957b72.png)

**Reply by EliasA97** [reply]
Well this is great. I was looking something like your design. Even the official datasheet of this module doesn't have a refrence schematic as explanatory as yours. I'll just stick with your schematic. The only thing I want anyway is to send some data from my nodes, so I don't need any special function

**Comment by EliasA97** [comment]
Hi Jan, just for educational reasons, how's that Q3 transistor will work?

**Reply by jgromes** [reply]
It's the simplest possible transistor inverter. When TXEN is low, the transistor doesn't conduct current, and si RXEN is pulled high by R12. When TXEN is high, the transistor conducts and pulls RXEN to ground.

**Reply by EliasA97** [reply]
And the TXen is controlled by DIO2 that's controlled by the library. I got it.
